### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by MarkOWiesemann** [uniuuu/zotprime#32] [General]
Hi uniuuu,

I am very happy to see that you keep on working on this project.
I was just wondering if you would mind me asking what the current roadmap is. Do I see it correctly that you are currently working on updating the repos to the upstreams?
And could I be of any help?

Best regards,
Mark

**Comment by uniuuu** [comment]
Hi @MarkOWiesemann 
Good day.
Plan is to check on issues, get the 7.0 client and introduce ARM. 
I'll count on you then. 
Meanwhile you may check on adding CA and https into docker compose setup.

Sincerely 
Alexander

**Reply by MarkOWiesemann** [reply]
Hi @uniuuu,
Good day to you too. I would love the 7.0 client too!
I can take a look at adding CA and https into docker compose setup. Do you have an issue or discussion you can point me to for more details? I found https://github.com/uniuuu/zotprime/discussions/17 but it seems to be a discussion about the localhost/VM setup rather about CA&HTTPS

Sincerely,
Mark

**Reply by uniuuu** [reply]
Hi @MarkOWiesemann 
You can try version 7 by building it with next Dockerfile: https://github.com/uniuuu/zotprime/blob/development/prebuild_client.Dockerfile 
I meantime could you please try to check on "preprint" error issue https://github.com/uniuuu/zotprime/discussions/15  ?

**Reply by MarkOWiesemann** [reply]
Hi @uniuuu ,

I build the version 7. I had some issues when using a local DNS name but as it worked when specifying the IP instead of DNS name I suspect it is an issue with my DNS setup. I might take a deeper look in case I want to verify that someday but for now I'd suggest, if you don't have the same problem, to leave it as is.

I had similar issues about unsupported itemTypes and will have a look at #15 .

**Reply by uniuuu** [reply]
Hi @MarkOWiesemann 
You may try now to build latest client (latest main branch commit)  with https://github.com/uniuuu/zotprime/blob/production/client.Dockerfile 
See more details for "preprint" error issue in next message https://github.com/uniuuu/zotprime/discussions/15#discussioncomment-7350612

**Reply by MarkOWiesemann** [reply]
Good evening @uniuuu,

I build the file you linked but got an error about checking out the tags:

```#11 [stage-1  7/14] RUN git checkout tags/7.0.6 -b v7.0.6
#11 0.190 fatal: 'tags/7.0.6' is not a commit and a branch 'v7.0.6' cannot be created from it
#11 ERROR: process "/bin/sh -c git checkout tags/7.0.6 -b v7.0.6" did not complete successfully: exit code: 128

#10 [stage-2  2/10] RUN set -eux;     apk update && apk upgrade &&     apk add --update --no-cache     file     git     git-lfs     grep     bash     coreutils     curl     ncurses     openssl     perl     python3     rsync     tar     xz     zip     7zip     && rm -rf /var/cache/apk/*
#10 0.206 fetch https://dl-cdn.alpinelinux.org/alpine/v3.20/main/x86_64/APKINDEX.tar.gz
#10 CANCELED
------
 > [stage-1  7/14] RUN git checkout tags/7.0.6 -b v7.0.6:
0.190 fatal: 'tags/7.0.6' is not a commit and a branch 'v7.0.6' cannot be created from it
------
client.Dockerfile:14
--------------------
  12 |     
  13 |     WORKDIR /usr/src/app/client/zotero-client
  14 | >>> RUN git checkout tags/7.0.6 -b v7.0.6
  15 |     #RUN git checkout 7.0.6-hotfix
  16 |     # RUN git checkout main
--------------------
ERROR: failed to solve: process "/bin/sh -c git checkout tags/7.0.6 -b v7.0.6" did not complete successfully: exit code: 128
```

**Reply by MarkOWiesemann** [reply]
> See more details for "preprint" error issue in next message [#15 (comment)](https://github.com/uniuuu/zotprime/discussions/15#discussioncomment-7350612)

I am currently having a look into it and trying to understand the connections from https://github.com/zotero/zotero-schema to the client and dataserver because if I understand it correctly both need to have the same db schema (one for checking locally if it is allowed and one for storing the related data)

To my understanding they started with the core data and switched to the defining the schema [here](https://github.com/zotero/zotero-schema/blob/master/schema.json).

**Reply by uniuuu** [reply]
> Good evening @uniuuu,
> 
> I build the file you linked but got an error about checking out the tags:
> 
> ```
> #11 0.190 fatal: 'tags/7.0.6' is not a commit and a branch 'v7.0.6' cannot be created from it
> #11 ERROR: process "/bin/sh -c git checkout tags/7.0.6 -b v7.0.6" did not complete successfully: exit code: 128
> 
> #10 [stage-2  2/10] RUN set -eux;     apk update && apk upgrade &&     apk add --update --no-cache     file     git     git-lfs     grep     bash     coreutils     curl     ncurses     openssl     perl     python3     rsync     tar     xz     zip     7zip     && rm -rf /var/cache/apk/*
> #10 0.206 fetch https://dl-cdn.alpinelinux.org/alpine/v3.20/main/x86_64/APKINDEX.tar.gz
> #10 CANCELED
> ------
>  > [stage-1  7/14] RUN git checkout tags/7.0.6 -b v7.0.6:
> 0.190 fatal: 'tags/7.0.6' is not a commit and a branch 'v7.0.6' cannot be created from it
> ------
> client.Dockerfile:14
> --------------------
>   12 |     
>   13 |     WORKDIR /usr/src/app/client/zotero-client
>   14 | >>> RUN git checkout tags/7.0.6 -b v7.0.6
>   15 |     #RUN git checkout 7.0.6-hotfix
>   16 |     # RUN git checkout main
> --------------------
> ERROR: failed to solve: process "/bin/sh -c git checkout tags/7.0.6 -b v7.0.6" did not complete successfully: exit code: 128
> ```

Hi @MarkOWiesemann 
It shows that submodule is outdated in your local copy of repository.
Try to clone repository again and use latest command from README.md

**Reply by MarkOWiesemann** [reply]
Hi @uniuuu,
yeah that worked. Sadly I am not very used to git submodules.

**Reply by uniuuu** [reply]
> Good evening @uniuuu,
> 
> I build the file you linked but got an error about checking out the tags:
> 
> ```
> #11 0.190 fatal: 'tags/7.0.6' is not a commit and a branch 'v7.0.6' cannot be created from it
> #11 ERROR: process "/bin/sh -c git checkout tags/7.0.6 -b v7.0.6" did not complete successfully: exit code: 128
> 
> #10 [stage-2  2/10] RUN set -eux;     apk update && apk upgrade &&     apk add --update --no-cache     file     git     git-lfs     grep     bash     coreutils     curl     ncurses     openssl     perl     python3     rsync     tar     xz     zip     7zip     && rm -rf /var/cache/apk/*
> #10 0.206 fetch https://dl-cdn.alpinelinux.org/alpine/v3.20/main/x86_64/APKINDEX.tar.gz
> #10 CANCELED
> ------
>  > [stage-1  7/14] RUN git checkout tags/7.0.6 -b v7.0.6:
> 0.190 fatal: 'tags/7.0.6' is not a commit and a branch 'v7.0.6' cannot be created from it
> ------
> client.Dockerfile:14
> --------------------
>   12 |     
>   13 |     WORKDIR /usr/src/app/client/zotero-client
>   14 | >>> RUN git checkout tags/7.0.6 -b v7.0.6
>   15 |     #RUN git checkout 7.0.6-hotfix
>   16 |     # RUN git checkout main
> --------------------
> ERROR: failed to solve: process "/bin/sh -c git checkout tags/7.0.6 -b v7.0.6" did not complete successfully: exit code: 128
> ```

Hi @MarkOWiesemann 
I was able to reproduce the error quoted above and patched it with latest fix https://github.com/uniuuu/zotprime/issues/37

**Comment by uniuuu** [comment]
Hi @MarkOWiesemann 
Thank you for resolving the 'preprint' issue!

There are more php files in the folder. 
![image](https://github.com/user-attachments/assets/e169167e-a29c-44ad-a812-77a67677361c)

Do you know which else we may need for the dataserver initialization ?

**Reply by MarkOWiesemann** [reply]
I can't really tell:
- delete_user does what it says
- move_library is a utility to move a library from one shard to another. I believe shards are some part of distributed database, so used to balance load
- purge_deleted_libraries, storage_existing_files -> storage_purge are cleanup scripts
- sanitize_notes does some kind of cleaning using https://github.com/zotero/dataserver/blob/master/model/Notes.inc.php#L85 and the Z_CONFIG::$HTMLCLEAN_SERVER_URL whatever that is. I don't understand why notes would not be cleaned when uploaded but idk.
- update_relations and update_sort_fields seem to be a utility function to balance load on shards

But that's all just from a quick glance so take it with a grain of salt.

**Comment by uniuuu** [comment]
@MarkOWiesemann There is a new challenge to find out whether it's possible to use Groups like it was done in previous versions of Zotero - see quoted repository where group and groupuser scripts are exist  https://github.com/uniuuu/zotprime/issues/27

**Reply by uniuuu** [reply]
@MarkOWiesemann Added group and user management scripts from https://github.com/victoradrianjimenez/dockerized-zotero/tree/master/dataserver/scripts  
Currently it's pending to fix SQL queries from these scripts as columns aren't match, i.e.:
`ERROR 1054 (42S22) at line 1: Unknown column 'role' in 'field list'`
 
Guide how to run them: https://github.com/uniuuu/zotprime/tree/development?tab=readme-ov-file#manage-users-and-groups

If you're available, please see if you can make a PR or share fix here. 
It's added in development branch https://github.com/uniuuu/zotprime/tree/development

Note there is php script available for adding group too in https://github.com/piernov/zotprime/blob/zotprime6/admin/addgroup

**Reply by MarkOWiesemann** [reply]
I will have a look soon

**Reply by uniuuu** [reply]
Hi @MarkOWiesemann 

> `ERROR 1054 (42S22) at line 1: Unknown column 'role' in 'field list'`

The init-mysql.sh is done differently here https://github.com/victoradrianjimenez/dockerized-zotero/blob/master/dataserver/init/init-mysql.sh comparing to used currently. 
And in www.sql there is the column "role":
`role enum('normal','deleted') NOT NULL DEFAULT 'normal',`  https://github.com/victoradrianjimenez/dockerized-zotero/blob/master/dataserver/init/www.sql#L28C1-L28C61  
comparing to the current one. 

Is it something you could look into ?

**Comment by uniuuu** [comment]
Hi @MarkOWiesemann 
For issue https://github.com/uniuuu/zotprime/issues/36
Please delete containers and volumes with `docker compose down -v`, remove repository and clone it again. As per updated README run ./bin/install.sh script [Dependencies and source code](https://github.com/uniuuu/zotprime?tab=readme-ov-file#dependencies-and-source-code)

**Reply by MarkOWiesemann** [reply]
Hi,
sorry for the late answer, I had a lot of stress starting my studies.
It all works now except annotations but I will make a clear report as soon as I get to it.

Never the less I wrote a liltle support article on an issue I got. Maybe you find a place to put it:
`If you're encountering the error _libraryStorageVersion cannot decrease in Zotero, you can resolve this by following these steps:

Reset Online Library:

	Go to Edit > Preferences > Sync.
	Under the Reset tab, choose Replace Online Library. This will reset the online library, but you might then encounter the error _libraryVersion cannot decrease.

Retrieve Current Version Numbers:

	Open the JavaScript console via Tools > Developer > Run JavaScript.
	Execute the following command:
	```
	Zotero.Libraries.userLibrary.libraryVersion + ' ' + Zotero.Libraries.userLibrary.storageVersion
	```
	Note down the two numbers that are printed (e.g., libraryVersion and storageVersion). These will be required later.

Access the Zotero Database:

	Connect to the database server hosting your Zotero data. If you’re using ZotPrime, access it via the zotprime-db container, typically on port 3306.
	Use a database client (e.g., MySQL Workbench, DBeaver, or the mysql CLI) to connect.

Update Version Numbers in the Database:

	Locate the appropriate database. For example, in a default setup, this might be zotero_shard_2.
	Within the database, navigate to the shardLibraries table.
	Update the version field for the relevant library to match the values retrieved earlier from the JavaScript console.

Resync:
	Restart Zotero and attempt to sync again. The process should now proceed without errors.

This process ensures that the versions in your database are consistent with Zotero's expectations, resolving the _libraryVersion cannot decrease error effectively.`

**Reply by uniuuu** [reply]
Hi @MarkOWiesemann 
Good day!
Do you know if it can be done through on of script in zotprime/stack/dataserver/dbconfig ?

**Comment by MarkOWiesemann** [comment]
By the way I just tried the windows arm client too. Works wonderful as well as syncing between my debian 64 laptop and the arm surface pro device. Congrats. Couldn't be more happy!

**Comment by uniuuu** [comment]
Hi @MarkOWiesemann 
Good day.
Group support has been added. ARM64 support added. 
Please clone latest code. Thank you!
